exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java

```diff
@@ -0,0 +1,543 @@
+  private static final ThrottlingLogger THROTTLING_LOGGER = new ThrottlingLogger(LOGGER);
+  private final boolean otelScopeEnabled;
+  private static final String OTEL_SCOPE_NAME = "otel_scope_name";
+  private static final String OTEL_SCOPE_VERSION = "otel_scope_version";
+  private static final long NANOS_PER_MILLISECOND = TimeUnit.MILLISECONDS.toNanos(1);
+
+  /**
+   * Constructor with feature flag parameter.
+   *
+   * @param otelScopeEnabled enable generation of the OpenTelemetry instrumentation scope info
+   *     metric and labels.
+   */
+  Otel2PrometheusConverter(boolean otelScopeEnabled) {
+    this.otelScopeEnabled = otelScopeEnabled;
+  }
+
+  MetricSnapshots convert(@Nullable Collection<MetricData> metricDataCollection) {
+    if (metricDataCollection == null || metricDataCollection.isEmpty()) {
+      return MetricSnapshots.of();
+    }
+    Map<String, MetricSnapshot> snapshotsByName = new HashMap<>(metricDataCollection.size());
+    Resource resource = null;
+    Set<InstrumentationScopeInfo> scopes = new LinkedHashSet<>();
+    for (MetricData metricData : metricDataCollection) {
+      MetricSnapshot snapshot = convert(metricData);
+      if (snapshot == null) {
+        continue;
+      }
+      putOrMerge(snapshotsByName, snapshot);
+      if (resource == null) {
+        resource = metricData.getResource();
+      }
+      if (otelScopeEnabled && !metricData.getInstrumentationScopeInfo().getAttributes().isEmpty()) {
+        scopes.add(metricData.getInstrumentationScopeInfo());
+      }
+    }
+    if (resource != null) {
+      putOrMerge(snapshotsByName, makeTargetInfo(resource));
+    }
+    if (otelScopeEnabled && !scopes.isEmpty()) {
+      putOrMerge(snapshotsByName, makeScopeInfo(scopes));
+    }
+    return new MetricSnapshots(snapshotsByName.values());
+  }
+
+  @Nullable
+  private MetricSnapshot convert(MetricData metricData) {
+    MetricMetadata metadata = convertMetadata(metricData);
+    InstrumentationScopeInfo scope = metricData.getInstrumentationScopeInfo();
+    switch (metricData.getType()) {
+      case LONG_GAUGE:
+        return convertLongGauge(metadata, scope, metricData.getLongGaugeData().getPoints());
+      case DOUBLE_GAUGE:
+        return convertDoubleGauge(metadata, scope, metricData.getDoubleGaugeData().getPoints());
+      case LONG_SUM:
+        SumData<LongPointData> longSumData = metricData.getLongSumData();
+        if (longSumData.getAggregationTemporality() == AggregationTemporality.DELTA) {
+          return null;
+        } else if (longSumData.isMonotonic()) {
+          return convertLongCounter(metadata, scope, longSumData.getPoints());
+        } else {
+          return convertLongGauge(metadata, scope, longSumData.getPoints());
+        }
+      case DOUBLE_SUM:
+        SumData<DoublePointData> doubleSumData = metricData.getDoubleSumData();
+        if (doubleSumData.getAggregationTemporality() == AggregationTemporality.DELTA) {
+          return null;
```

Comment:
Maybe add a comment on these delta cases that they should never happen because `PrometheusMetricReader#getAggregationTemporality` returns `CUMULATIVE`.

Reply:
done